habitat_baselines/rl/ver/preemption_decider.py

```diff
@@ -0,0 +1,403 @@
+
+        if target_length_time > 0.0:
+            self.opt_rollout_time += target_length_time
+            step_times = my_step_averages[:, np.newaxis] * np.arange(
+                1, max_possible_steps + 1, dtype=np.float64
+            )
+
+            self.my_opt_rollout_steps = np.count_nonzero(
+                step_times <= float(self.opt_rollout_time)
+            )
+
+        if (
+            self._learning_time.count == self._learning_time.window_size
+            and self.opt_rollout_time.count
+            == self.opt_rollout_time.window_size
+        ):
+            self.rollout_ends.steps.value = float(self.my_opt_rollout_steps)
+        else:
+            self.rollout_ends.steps.value = -1
+
+    def policy_step(self, data):
+        step_time = data["t_stamp"]
+        for _, env_idx in data["steps_finished"]:
+            if self.last_step_times[env_idx] > 0:
+                self.step_averages[env_idx] += (
+                    step_time - self.last_step_times[env_idx]
+                )
+
+            self.last_step_times[env_idx] = step_time
+
+        self.real_steps_collected += len(data["steps_finished"])
+
+    def start_rollout(self, start_time):
+        self.start_time = self._all_reduce(
+            (start_time - self.my_t_zero),
+            reduce_op=torch.distributed.ReduceOp.MIN,
+        )
+
+        self.n_rollouts_started += 1
+        self.last_step_times[:] = -1.0
+
+        self.started = True
+        if (
+            self._learning_time.count == self._learning_time.window_size
+            and self.opt_rollout_time.count
+            == self.opt_rollout_time.window_size
+        ):
+            self.rollout_ends.time.value = (
+                self.my_t_zero + self.start_time + float(self.opt_rollout_time)
+            )
+        else:
+            self.rollout_ends.time.value = -1
+
+    def end_rollout(self, end_steps_time, num_next_steps):
+        end_steps_time = self._all_reduce(
+            (end_steps_time - self.my_t_zero),
+            reduce_op=torch.distributed.ReduceOp.MAX,
+        )
+        self.queues.report.put(
+            (
+                ReportWorkerTasks.preemption_decider,
+                dict(
+                    real_steps_collected=self.real_steps_collected,
+                    expected_steps_collected=self.expected_steps_collected,
+                    real_rollout_time=(end_steps_time - self.start_time) * 1e3,
+                    expected_rollout_time=float(self.opt_rollout_time) * 1e3,
+                ),
+            )
+        )
+
+        if self.rollout_ends.time.value > 0:
+            self.preemption_error_time += (
+                end_steps_time - self.start_time
+            ) - float(self.opt_rollout_time)
+
+        self.started = False
+        self.rollout_ends.time.value = -1.0
+        self.rollout_ends.steps.value = -1.0
+
+        self.update(num_next_steps)
+        self.real_steps_collected = 0
+
+    def learner_time(self, learner_time):
```

Comment:
Some typehints to the methods here would be useful to understand what's going on.